### DISCUSSION TITLE
응용 서비스 계층과 인프라계층의 의존 관계

### DISCUSSION THREAD

**Discussion by yxxnghwan** [woowacourse-teams/2022-nae-pyeon#233] [BE 컨벤션 이슈]
프로젝트 초기에 응용 서비스 계층이 Spring Data JPA에 의존적이지 않게 짜자! 라는 결론을 짓고 그 해결방법으로 Repository계층을 인터페이스로 정의했었어요. 그런데 이후 Spring Data Jpa에서 제공하는 Page, Pageable API들을 활용할 일이 생겼는데, 이것들을 기존의 의존관계를 생각하며 사용하려면 응용 서비스에 새로운 DTO를 정의하고 해당 DTO로 포장해서 데이터를 전달해야하는 설계 복잡도가 추가로 올라가더라구요.
해당 계층이 존재함으로써 다른 인프라 기술로 변경시 변경의 유연함은 존재하고 응용 서비스 계층을 보호할 수 있지만, 설계의 복잡함으로 인한 개발 피로도 및 생산성에 저하가 있는 것 같아요!
해당 설계에 대한 의견을 한번씩 더 들어보고 싶네요!

**Comment by kth990303** [comment]
지금의 제 생각은 Spring Data JPA에 의존적인 Repository 계층을 사용해도 괜찮은 듯해요. 
특정 기술을 사용하기로 한 순간부터 어느정도의 의존은 불가피하다고 생각되며, Spring Data JPA의 장점인 반복을 줄이고 편리한 db 로직 관리를 버릴 만큼, 먼 미래의 기술 변경의 유연함을 챙길 필요는 없다고 생각되기 때문이에요. 또한, Spring Data JPA에 의존하지 않도록 인터페이스를 만들어 구조를 짜면, 해당 설계에 대한 학습 장벽이 높아지고 불필요한 반복이 증가해 오히려 개발자의 피로감을 높인다는 생각도 듭니다 😢 
스프린트4 때는 계층을 추상화하지 않는 방법을 사용해도 괜찮다고 생각되네요!
